File: go/internal/ui/badge.go
package ui

import (
	"fmt"
	"strings"
)

// BadgeVariant represents the possible variants of a badge
type BadgeVariant string

const (
	// BadgeVariantMuted represents a muted badge
	BadgeVariantMuted BadgeVariant = "muted"
	// BadgeVariantNeutral represents a neutral badge
	BadgeVariantNeutral BadgeVariant = "neutral"
	// BadgeVariantInfo represents an info badge
	BadgeVariantInfo BadgeVariant = "info"
	// BadgeVariantSuccess represents a success badge
	BadgeVariantSuccess BadgeVariant = "success"
	// BadgeVariantWarning represents a warning badge
	BadgeVariantWarning BadgeVariant = "warning"
	// BadgeVariantDanger represents a danger badge
	BadgeVariantDanger BadgeVariant = "danger"
	// BadgeVariantTier represents a tier badge
	BadgeVariantTier BadgeVariant = "tier"
)

// Badge represents a badge component
type Badge struct {
	Text         string
	Icon         string
	IconClasses  []string
	IconOnly     bool
	Href         string
	Variant      BadgeVariant
	HTMLOptions  map[string]interface{}
	Content      string
}

// NewBadge creates a new badge with the given parameters
func NewBadge(text string, icon string, iconClasses []string, iconOnly bool, href string, variant BadgeVariant, htmlOptions map[string]interface{}, content string) *Badge {
	if iconClasses == nil {
		iconClasses = []string{}
	}
	if htmlOptions == nil {
		htmlOptions = make(map[string]interface{})
	}

	return &Badge{
		Text:         text,
		Icon:         icon,
		IconClasses:  iconClasses,
		IconOnly:     iconOnly,
		Href:         href,
		Variant:      variant,
		HTMLOptions:  htmlOptions,
		Content:      content,
	}
}

// BadgeClasses returns the CSS classes for the badge
func (b *Badge) BadgeClasses() string {
	var classes []string

	// Add base classes
	classes = append(classes, "gl-badge", "badge", "badge-pill", fmt.Sprintf("badge-%s", b.Variant))

	// Add icon-only class if needed
	if b.IconOnly {
		classes = append(classes, "!gl-px-2")
	}

	return strings.Join(classes, " ")
}

// IconClasses returns the CSS classes for the badge icon
func (b *Badge) IconClasses() string {
	var classes []string

	// Add base classes
	classes = append(classes, "gl-icon", "gl-badge-icon")
	classes = append(classes, b.IconClasses...)

	// Add margin class if needed
	if b.CircularIcon() {
		classes = append(classes, "-gl-ml-2")
	}

	return strings.Join(classes, " ")
}

// IconOnly returns whether the badge should only show the icon
func (b *Badge) IconOnly() bool {
	return b.IconOnly
}

// Link returns whether the badge should be rendered as a link
func (b *Badge) Link() bool {
	return b.Href != ""
}

// Text returns the text content of the badge
func (b *Badge) Text() string {
	if b.Content != "" {
		return b.Content
	}
	return b.Text
}

// HasIcon returns whether the badge has an icon
func (b *Badge) HasIcon() bool {
	return b.IconOnly() || b.Icon != ""
}

// CircularIcon returns whether the icon should be circular
func (b *Badge) CircularIcon() bool {
	return b.Icon == "issue-open-m" || b.Icon == "issue-close"
}

// HTMLOptions returns the HTML options for the badge
func (b *Badge) HTMLOptions() map[string]interface{} {
	options := make(map[string]interface{})
	for k, v := range b.HTMLOptions {
		options[k] = v
	}

	// Add CSS classes
	options["class"] = b.BadgeClasses()

	// Add aria attributes for icon-only badges
	if b.IconOnly() {
		if _, ok := options["aria"]; !ok {
			options["aria"] = make(map[string]interface{})
		}
		aria := options["aria"].(map[string]interface{})
		aria["label"] = b.Text()
		options["role"] = "img"
	}

	return options
}

// Render generates the HTML for the badge
func (b *Badge) Render() string {
	// Build HTML options
	htmlOptions := b.HTMLOptions()

	// Determine the tag to use
	tag := "span"
	if b.Link() {
		tag = "a"
		htmlOptions["href"] = b.Href
	}

	// Build the content
	var contentParts []string

	// Add icon if needed
	if b.HasIcon() {
		contentParts = append(contentParts, fmt.Sprintf(`<span class="%s">%s</span>`, b.IconClasses(), b.Icon))
	}

	// Add text if needed
	if text := b.Text(); text != "" {
		contentParts = append(contentParts, fmt.Sprintf(`<span class="gl-badge-content">%s</span>`, text))
	}

	// Render the badge
	return fmt.Sprintf(`<%s %s>%s</%s>`,
		tag,
		formatAttributes(htmlOptions),
		strings.Join(contentParts, ""),
		tag)
}

File: rust_app/src/components/pajamas/badge.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Variant options for the badge component
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BadgeVariant {
    Muted,
    Neutral,
    Info,
    Success,
    Warning,
    Danger,
    Tier,
}

impl BadgeVariant {
    fn as_str(&self) -> &'static str {
        match self {
            BadgeVariant::Muted => "muted",
            BadgeVariant::Neutral => "neutral",
            BadgeVariant::Info => "info",
            BadgeVariant::Success => "success",
            BadgeVariant::Warning => "warning",
            BadgeVariant::Danger => "danger",
            BadgeVariant::Tier => "tier",
        }
    }
}

/// Renders a Pajamas compliant badge element
#[derive(Debug, Clone)]
pub struct BadgeComponent {
    /// The text content of the badge
    text: Option<String>,
    /// The icon name
    icon: Option<String>,
    /// Additional CSS classes for the icon
    icon_classes: Vec<String>,
    /// Whether to show only the icon
    icon_only: bool,
    /// The href for the badge link
    href: Option<String>,
    /// The variant of the badge
    variant: BadgeVariant,
    /// Additional HTML options
    html_options: HashMap<String, String>,
    /// The content slot
    content: Option<String>,
}

impl BadgeComponent {
    /// Create a new badge component
    pub fn new(
        text: Option<String>,
        icon: Option<String>,
        icon_classes: Option<Vec<String>>,
        icon_only: bool,
        href: Option<String>,
        variant: BadgeVariant,
        html_options: Option<HashMap<String, String>>,
    ) -> Self {
        Self {
            text,
            icon,
            icon_classes: icon_classes.unwrap_or_default(),
            icon_only,
            href,
            variant,
            html_options: html_options.unwrap_or_default(),
            content: None,
        }
    }

    /// Set the content slot
    pub fn with_content(mut self, content: String) -> Self {
        self.content = Some(content);
        self
    }

    /// Get the badge classes
    fn badge_classes(&self) -> String {
        let mut classes = vec![
            "gl-badge".to_string(),
            "badge".to_string(),
            "badge-pill".to_string(),
            format!("badge-{}", self.variant.as_str()),
        ];

        if self.icon_only {
            classes.push("!gl-px-2".to_string());
        }

        classes.join(" ")
    }

    /// Get the icon classes
    fn icon_classes(&self) -> String {
        let mut classes = vec!["gl-icon".to_string(), "gl-badge-icon".to_string()];
        classes.extend(self.icon_classes.clone());

        if self.circular_icon() {
            classes.push("-gl-ml-2".to_string());
        }

        classes.join(" ")
    }

    /// Check if the badge is icon-only
    fn icon_only(&self) -> bool {
        self.icon_only
    }

    /// Check if the badge is a link
    fn is_link(&self) -> bool {
        self.href.is_some()
    }

    /// Get the text content
    fn text(&self) -> Option<String> {
        self.content.clone().or_else(|| self.text.clone())
    }

    /// Check if the badge has an icon
    fn has_icon(&self) -> bool {
        self.icon_only || self.icon.is_some()
    }

    /// Check if the icon is circular
    fn circular_icon(&self) -> bool {
        if let Some(icon) = &self.icon {
            matches!(icon.as_str(), "issue-open-m" | "issue-close")
        } else {
            false
        }
    }

    /// Get the HTML options
    fn html_options(&self) -> HashMap<String, String> {
        let mut options = self.html_options.clone();
        options.insert("class".to_string(), self.badge_classes());

        if self.icon_only {
            if let Some(text) = self.text() {
                options.insert("aria-label".to_string(), text);
                options.insert("role".to_string(), "img".to_string());
            }
        }

        options
    }

    /// Render the badge component
    pub fn render(&self) -> String {
        let mut html = String::new();
        let options = self.html_options();
        let options_str = options
            .iter()
            .map(|(key, value)| format!(" {}=\"{}\"", key, value))
            .collect::<Vec<String>>()
            .join("");

        if self.is_link() {
            html.push_str(&format!(
                r#"<a href="{}"{}>"#,
                self.href.as_ref().unwrap(),
                options_str
            ));
        } else {
            html.push_str(&format!("<span{}>", options_str));
        }

        if self.has_icon() {
            html.push_str(&format!(
                r#"<svg class="sprite-icon {}" aria-hidden="true"><use xlink:href="/assets/icons.svg#{}"></use></svg>"#,
                self.icon_classes(),
                self.icon.as_ref().unwrap()
            ));
        }

        if let Some(text) = self.text() {
            html.push_str(&format!(
                r#"<span class="gl-badge-content">{}</span>"#,
                text
            ));
        }

        if self.is_link() {
            html.push_str("</a>");
        } else {
            html.push_str("</span>");
        }

        html
    }
}
